async/library_async_nativeiofs.js

```diff
@@ -0,0 +1,372 @@
+      return -{{{cDefine('ENOSYS')}}};
+    },
+
+    fcntl: async function(fd, cmd) {
+      if( cmd != {{{ cDefine('F_SETLK') }}} ) {
+        return -{{{cDefine('ENOSYS')}}};
+      }
+      return 0;// Pretend that the locking was successful.
+    },
+
+    read: async function(fd, buffer, offset, length) {
+      var fh = AsyncFSImpl.fileDescriptorToFileHandle[fd];
+      var data = buffer.subarray(offset, offset + length);
+      let bytes_read = await fh.read(data, fh.seek_position);
+      fh.seek_position += bytes_read;
+      buffer.set(data, offset);
+      return bytes_read;
+    },
+
+    write: async function(fd, buffer, offset, length, position) {
+      if (length < 0 || position < 0) {
+        throw new FS.ErrnoError({{{ cDefine('EINVAL') }}});
+      }
+      var fh = AsyncFSImpl.fileDescriptorToFileHandle[fd];
+      var seeking = typeof position !== 'undefined';
+      if (!seeking) {
+        position = fh.seek_position;
+      }
+      var data = buffer.subarray(offset, offset + length);
+      let bytes_written = await fh.write(data, position);
+      fh.timestamp = Date.now();
+      if (!seeking) fh.seek_position += bytes_written;
+      return bytes_written;
+    },
+
+    readlink: async function(path, buf, bufsize) {
+      return -{{{cDefine('ENOSYS')}}};
+    },
+
+    munmap: async function(addr, len) {
+      return -{{{cDefine('ENOSYS')}}};
+    },
+
+    fchmod: async function(fd, mode) {
+      return -{{{cDefine('ENOSYS')}}};
+    },
+
+    fsync: async function(fd) {
+      await AsyncFSImpl.fileDescriptorToFileHandle[fd].flush();
+      return 0;
+    },
+
+    mmap2: async function(addr, len, prot, flags, fd, off) {
+      return -{{{cDefine('ENOSYS')}}};
+    },
+
+    readv: async function(fd, iovs) {
+      var ret = 0;
+      for (var i = 0; i < iovcnt; i++) {
+        var ptr = {{{ makeGetValue('iov', 'i*8', 'i32') }}};
+        var len = {{{ makeGetValue('iov', 'i*8 + 4', 'i32') }}};
+        var curr = await AsyncFSImpl.read(fd, {{{ heapAndOffset('HEAP8', 'ptr') }}}, len, offset);
+        if (curr < 0) return -1;
+        ret += curr;
+        if (curr < len) break; // nothing more to read
+      }
+      return ret;
+    },
+
+    writev: async function(fd, iov, iovcnt, offset) {
+      var ret = 0;
+      for (var i = 0; i < iovcnt; i++) {
+        var ptr = {{{ makeGetValue('iov', 'i*8', 'i32') }}};
+        var len = {{{ makeGetValue('iov', 'i*8 + 4', 'i32') }}};
+        var curr = await AsyncFSImpl.write(fd, {{{ heapAndOffset('HEAP8', 'ptr') }}}, len, offset);
+        if (curr < 0) return -1;
+        ret += curr;
+      }
+      return ret;
+    },
+
+    unlink: async function(pathname) {
+      var absolutePath = AsyncFSImpl.getAbsolutePath(pathname);
+
+      encodedPath = AsyncFSImpl.encodePath(absolutePath);
+      try {
+        await nativeIO.delete(encodedPath);
+        return 0;
+      } catch(e) {
+        console.log("Unlink faild with error", e);
+        return e;
```

Comment:
Should you return an error code instead of e here?

Reply:
Done